pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -3942,9 +3942,13 @@ public void publishTxnMessage(TxnID txnID, ByteBuf headersAndPayload, PublishCon
     @Override
     public CompletableFuture<Void> endTxn(TxnID txnID, int txnAction, long lowWaterMark) {
         if (TxnAction.COMMIT_VALUE == txnAction) {
-            return transactionBuffer.commitTxn(txnID, lowWaterMark);
+            return transactionBuffer.commitTxn(txnID, lowWaterMark).thenRun(() -> {
+                lastDataMessagePublishedTimestamp = Clock.systemUTC().millis();
+            });
         } else if (TxnAction.ABORT_VALUE == txnAction) {
-            return transactionBuffer.abortTxn(txnID, lowWaterMark);
+            return transactionBuffer.abortTxn(txnID, lowWaterMark).thenRun(() -> {
+                lastDataMessagePublishedTimestamp = Clock.systemUTC().millis();
```

Comment:
need to consider race conditions

Reply:
Do we need to guarantee strict correctess?

Reply:
I'm afraid yes, to prevent potential issues.
Just ```java
transactionBuffer.abortTxn(txnId, lowWaterMarker).thenRunAsync(() -> {
 .... 
}, ((ManagedLedgerImpl) ledger).getExecutor());
```
is OK, make update `lastDataMessagePublishedTimestamp ` in a same thread.

Reply:
why here have a race conditions? @dao-jun

Reply:
> why here have a race conditions? @dao-jun

@Technoboy-  Oh, sorry. I didn't check the code, I thought the Future completed by the other thread, but in actually, they are in the same thread.